### PR TITLE
JavaScript use cork API, the 2nd trial

### PR DESCRIPTION
**Pull request by masatake**
This is a revised version of #2115.

JavaScript parser uses vStrings for tracking and building scope information.
This pull request replaces them with cork index integers.

About corkIndex, see http://docs.ctags.io/en/latest/internal.html?highlight=cork#cork-api.

In addition to using corkQueue, this change uses corkSymbolTable.
There is no document for corkSymbolTable. I must document it.

Not a few test cases are failed. More work is needed.

- [x] with this change, scope fields are filled with non-full-qualified names. This must be fixed.
- [x] revise which kinds should have chain-elements role.
- [x] some test cases must be updated because (1) bugs of the original code are exposed with this new cork-based code, and (2) this code introduces incompatible changes.  See also #2126 and ##3363.
- [x] Units/parser-javascript.r/jsFunc_tutorial.js.d
- [x] Units/parser-javascript.r/bug3036476.js.d
- [x] Units/review-needed.r/simple.js.t
- Add comments and documentation
  - [x] write about the cork symbol table (=> docs/).
  - [x] add comments to moveChildren().
  - [x] add comments to makeJsRefTagsForNameChain().
  - [X] add comments to "chain-element" role.
  
I will revise the code for OpenUI5 in another pull request.

This change is related to #2123.

### CODE REVIEW

**Comment by masatake** [comment]
@[jafl](https://github.com/jafl), as I wrote I applied the cork API to the JavaScript parser.
Though I must add more comments to the code I changed, I think it is ready for your evaluation.
What I would like you to know is that the vString-based code for tracking scope is gone.
I strongly believe cork index (integer) based code is better.

Though the code becomes a bit complicated. However, when we support import/export statements in the parser, I think the complexity will pay for itself.

**Comment by masatake** [comment]
@jafl, I added comments to my changes.

**Comment by jafl** [comment]
Thanks!  I will try to find time to review it.

**Comment by masatake** [comment]
Can I rebase this branch on the latest master branch?

**Comment by jafl** [comment]
Sure

**Comment by jafl** [comment]
I will study it today.  Sorry for the long delay.

File: parsers/jscript.c [hunk 1]
```diff
@@ -299,137 +395,197 @@ static void injectDynamicName (tokenInfo *const token, vString *newName)
...
+			return index;
+	}
+
+	tagEntryInfo e;
+	initTagEntry (&e, name, kind);
+	e.lineNumber   = token->lineNumber;
+	e.filePosition = token->filePosition;
+	e.extensionFields.scopeIndex = scope;
+
+#if 0
```
Comment (jafl) [inline parsers/jscript.c:526]: Should this be `#if DO_TRACING`?

File: parsers/jscript.c [hunk 2]
```diff
@@ -299,137 +395,197 @@ static void injectDynamicName (tokenInfo *const token, vString *newName)
...
+	int scope = token->scope;
+	if (name_chain)
+	{
+		scope = makeJsRefTagsForNameChain (name_chain, token, kind, scope);
+		eFree (name_chain);
+	}
+
+	/*
+	 * Check whether NAME is already defined in SCOPE.
+	 * If the NAME is already defined, return the cord index for the NAME.
```
Comment (jafl) [inline parsers/jscript.c:511]: "cord" -> "cork"

**Review by jafl — APPROVED** [review]
2 very minor points

**Comment by masatake** [comment]
Thank you for reviewing. I updated the pull request.

**Comment by jafl** [comment]
Once you merge this, I will start working on redoing my PR based on the new code.

**Comment by masatake** [comment]
I'll squash the last 3 commits into one before merging.

**Pull request merged**